src/cascadia/PublicTerminalCore/HwndTerminalAutomationPeer.cpp

```diff
@@ -0,0 +1,159 @@
+    // so we only really need to touch it for the properties we want to implement
+    if (propertyId == UIA_ClassNamePropertyId)
+    {
+        // IMPORTANT: Do NOT change the name. Screen readers like may be dependent on this being "WpfTermControl".
+        pVariant->bstrVal = SysAllocString(L"WPFTermControl");
+        if (pVariant->bstrVal != nullptr)
+        {
+            pVariant->vt = VT_BSTR;
+        }
+    }
+    else
+    {
+        // fall back to shared implementation
+        TermControlUiaProvider::GetPropertyValue(propertyId, pVariant);
+    }
+    return S_OK;
+}
+
+// Method Description:
+// - Signals the ui automation client that the terminal's selection has changed and should be updated
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HwndTerminalAutomationPeer::SignalSelectionChanged()
+{
+    UiaTracing::Signal::SelectionChanged();
+    LOG_IF_FAILED(UiaRaiseAutomationEvent(this, UIA_Text_TextSelectionChangedEventId));
+}
+
+// Method Description:
+// - Signals the ui automation client that the terminal's output has changed and should be updated
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HwndTerminalAutomationPeer::SignalTextChanged()
+{
+    UiaTracing::Signal::TextChanged();
+    LOG_IF_FAILED(UiaRaiseAutomationEvent(this, UIA_Text_TextChangedEventId));
+}
+
+// Method Description:
+// - Signals the ui automation client that the cursor's state has changed and should be updated
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HwndTerminalAutomationPeer::SignalCursorChanged()
+{
+    UiaTracing::Signal::CursorChanged();
+    LOG_IF_FAILED(UiaRaiseAutomationEvent(this, UIA_Text_TextSelectionChangedEventId));
+}
+
+void HwndTerminalAutomationPeer::NotifyNewOutput(std::wstring_view newOutput)
+{
+    // Try to suppress any events (or event data)
+    // that is just the keypress the user made
+    auto sanitized{ Sanitize(newOutput) };
+    while (!_keyEvents.empty() && IsReadable(sanitized))
+    {
+        if (til::toupper_ascii(sanitized.front()) == _keyEvents.front())
+        {
+            // the key event's character (i.e. the "A" key) matches
+            // the output character (i.e. "a" or "A" text).
+            // We can assume that the output character resulted from
+            // the pressed key, so we can ignore it.
+            sanitized = sanitized.substr(1);
+            _keyEvents.pop_front();
+        }
+        else
+        {
+            // The output doesn't match,
+            // so clear the input stack and
+            // move on to fire the event.
+            _keyEvents.clear();
+            break;
+        }
+    }
+
+    // Suppress event if the remaining text is not readable
+    if (!IsReadable(sanitized))
+    {
+        return;
+    }
+
+    const auto sanitizedBstr = SysAllocString(sanitized.c_str());
+    static auto activityId = SysAllocString(L"TerminalTextOutput");

```

Comment:
we might be leaking these -- can you confirm whether UiaRaise takes ownership? I bet it doesn't.

Should use `wil::unique_bstr`

Reply:
i believe the ones in the Variants ARE accounted for, and you do not need to delete them or use unique_bstr on them.